adapters/tradplus/tradplus.go

```diff
@@ -0,0 +1,176 @@
+package tradplus
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/macros"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint *template.Template
+}
+
+// Builder builds a new instance of the tradplus adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	return &adapter{
+		endpoint: template,
+	}, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	adapterRequest, errs := a.makeRequest(request)
+	if errs != nil {
+		return nil, errs
+	}
+	return []*adapters.RequestData{adapterRequest}, nil
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, []error) {
+
+	tradplusExt, err := getImpressionExt(&request.Imp[0])
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	request.Imp[0].Ext = nil
+
+	url, err := a.buildEndpointURL(tradplusExt)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	err = transform(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Body:    reqBody,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}, nil
+}
+
+func getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpTradPlus, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error parsing tradplusExt - " + err.Error(),
+		}
+	}
+
+	var tradplusExt openrtb_ext.ExtImpTradPlus
+	if err := json.Unmarshal(bidderExt.Bidder, &tradplusExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error parsing bidderExt - " + err.Error(),
+		}
+	}
+
+	if tradplusExt.AccountID == "" {
+		return nil, &errortypes.BadInput{
+			Message: "imp.ext.accountId required",
+		}
+	}
+
+	return &tradplusExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtImpTradPlus) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		AccountID: params.AccountID,
+		ZoneID:    params.ZoneID,
+	}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func transform(request *openrtb2.BidRequest) error {
```

Comment:
Recommend to choose a more specific name. such as transformToNative11 or something similar to make the functionality clear.

Reply:
This function has been removed, and supported in our service.